File: Extensions/Tessa.Extensions.Default.Shared/Workflow/KrProcess/Formatters/ApprovalStageTypeFormatter.cs
#nullable enable

using System.Collections.Generic;
using System.Threading.Tasks;
using Tessa.Platform;
using Tessa.Platform.Storage;

namespace Tessa.Extensions.Default.Shared.Workflow.KrProcess.Formatters
{
    /// <summary>
    /// Форматтер типа этапа <see cref="StageTypeDescriptors.ApprovalDescriptor"/>.
    /// </summary>
    public sealed class ApprovalStageTypeFormatter :
        StageTypeFormatterBase
    {
        #region Base Overrides

        /// <inheritdoc />
        public override async ValueTask FormatClientAsync(
            IStageTypeFormatterContext context)
        {
            await base.FormatClientAsync(context);

            FormatInternal(
                context,
                context.StageRow);
        }

        /// <inheritdoc />
        public override async ValueTask FormatServerAsync(
            IStageTypeFormatterContext context)
        {
            await base.FormatServerAsync(context);

            FormatInternal(
                context,
                context.Settings!);
        }

        #endregion

        #region Private Methods

        private static void FormatInternal(
            IStageTypeFormatterContext context,
            IDictionary<string, object?> storage)
        {
            var sb = StringBuilderHelper.Acquire(128);

            if (storage.TryGet<bool>(KrConstants.KrApprovalSettingsVirtual.Advisory))
            {
                sb.AppendLine("{$UI_KrApproval_Advisory}");
            }

            sb.Append(storage.TryGet<bool>(KrConstants.KrApprovalSettingsVirtual.IsParallel)
                ? "{$UI_KrApproval_Parallel}"
                : "{$UI_KrApproval_Sequential}");

            context.DisplaySettings = sb.ToStringAndRelease();
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workflow/krProcess/formatters/approvalStageTypeFormatter.ts
import { extension } from '@tessa/application';
import {
  IKrStageTypeFormatterContext,
  KrStageTypeFormatter,
  KrStageTypeFormattingHelper,
  StageTypeDescriptor,
  StageTypeDescriptors
} from '@tessa/platform';

@extension({ name: 'ApprovalStageTypeFormatter' })
export class ApprovalStageTypeFormatter extends KrStageTypeFormatter {
  //#region constants and static fields

  private static readonly _isAdvisory = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrApprovalSettingsVirtual',
    'Advisory'
  );
  private static readonly _isParallel = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrApprovalSettingsVirtual',
    'IsParallel'
  );

  //#endregion

  //#region base overrides

  descriptors(): ReadonlyArray<StageTypeDescriptor> {
    return [StageTypeDescriptors.approvalDescriptor];
  }

  format(context: IKrStageTypeFormatterContext): void {
    super.format(context);

    let sb = '';
    if (context.stageRow.tryGet(ApprovalStageTypeFormatter._isAdvisory)) {
      sb += '{$UI_KrApproval_Advisory}\n\r';
    }

    sb += context.stageRow.tryGet(ApprovalStageTypeFormatter._isParallel)
      ? '{$UI_KrApproval_Parallel}'
      : '{$UI_KrApproval_Sequential}';
    context.displaySettings = sb;
  }

  //#endregion
}
